### PR TITLE
[msal-node] samples and docs for using certificates, key vault, managed id

### LINKED ISSUE #3023: How to use msal-node with Managed Identity?
## Library

- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [ ] `@azure/msal-browser@2.x.x`
- [x] `@azure/msal-node@1.x.x`
- [ ] `@azure/msal-react@1.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angular@2.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Description

How can I use msal-node with Managed Identity?

I’m co-maintaining CLI for Microsoft 365 where we support a number of ways for users to authenticate. We’re looking into migrating from adal-node to msal-node but couldn’t find any examples of using Managed Identity in conjunction with MSAL token cache.

## Source

- [ ] Internal (Microsoft)
- [x] Customer request

### PR DESCRIPTION
**Pull request by derisen**
This PR adds 2 new samples and 2 corresponding docs

* **using certificates**
    * `auth-code-with-certs/`: demonstrates a web app using certificates and running an https server
    * `certificate-credentials.md`: explains creating certificates and initializing msal-node with them
* **using key vault & managed id**
    * `auth-code-key-vault/`: demonstrates a web app using Azure SDKs to access key vault for fetching certificates
    * `key-vault-managed-identity.md`: explains creating a key vault, uploading certificates, deploying a web app to use managed id

cc @mmacy

### CODE REVIEW

**Comment by derisen** [comment]
@sameerag @jasonnutter @mmacy this is now ready for review

**Review by samuelkubai — APPROVED** [review]
I dropped a question, otherwise this looks good to me.

File: lib/msal-node/docs/certificate-credentials.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+### Generating self-signed certificates
+
+Build and install **OpenSSL** for your **OS** following the guide at [github.com/openssl](https://github.com/openssl/openssl#build-and-install). If you like to skip building and get a binary distributable from the community instead, check the [OpenSSL Wiki: Binaries](https://wiki.openssl.org/index.php/Binaries) page.
+
+Afterwards, add the path to `OpenSSL` to your **environment variables** so that you can call it from anywhere.
+
+Type the following in a terminal. You will be prompted to enter a *pass phrase* (this is optional, but recommended):
+
+```bash
```
[the remark this answers is not in the document]
Comment (derisen) [reply lib/msal-node/docs/certificate-credentials.md:23]: Ah, didn't run into this myself. Let me check again

[the remark this answers is not in the document]
Comment (derisen) [reply lib/msal-node/docs/certificate-credentials.md:23]: Aah, looks like you were using it like:

> openssl req -x509 -newkey rsa:2048 -sha256 -days 365 -keyout example.key -out example.crt -subj "C:/Program Files/Git/CN=example.com"

`CN=example.com` here is not a file name, it's the "common name" parameter for the certificate, so you should not pass a file path there.

File: lib/msal-node/docs/key-vault-managed-identity.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+function msalApp(thumbprint, privateKey) {
+
+    // Before running the sample, you will need to replace the values in the config
+    const config = {
+        auth: {
+            clientId: "ENTER_CLIENT_ID",
+            authority: "https://login.microsoftonline.com/ENTER_TENANT_ID",
+            clientCertificate: {
+                thumbprint: thumbprint,
+                privateKey: privateKey,
```
Comment (sadasant) [inline lib/msal-node/docs/key-vault-managed-identity.md:80]: Hello team, sorry to chime in. Two questions:
- Is there a way to pass a non-PEM privateKey here?
- Is there a way to specify a password for the PEM certificate here?

Comment (derisen) [reply lib/msal-node/docs/key-vault-managed-identity.md:80]: @sadasant sorry for late answer.
1) No, the privateKey should be PEM encoded
2) No again, but you can use the [crypto module](https://nodejs.org/docs/latest-v14.x/api/crypto.html#crypto_crypto_createprivatekey_key) to decrypt the password protected privateKey before passing it to msal-node:

```javascript
const fs = require('fs');
const crypto = require('crypto');
const privateKeySource = fs.readFileSync('./example.key')
const privateKeyObject = crypto.createPrivateKey({
    key: privateKeySource,
    passphrase: "YOUR_PASSPHRASE",
    format: 'pem'
});
const privateKey = privateKeyObject.export({
    format: 'pem',
    type: 'pkcs8'
});
```

cc @sameerag

**Comment by derisen** [comment]
@sameerag any idea why the tests are failing?

**Comment by sameerag** [comment]
> @sameerag any idea why the tests are failing?

No Idea. I see this happening on other builds too, let me investigate and get back.

**Comment by sameerag** [comment]
> > @sameerag any idea why the tests are failing?
> 
> No Idea. I see this happening on other builds too, let me investigate and get back.

@derisen Seems to be a PPE environment issue.They have a ICM open which isn't addressed yet.

File: lib/msal-node/docs/certificate-credentials.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (samuelkubai) [reply lib/msal-node/docs/certificate-credentials.md:23]: I didn't pass the parameter as a file name, I used the command as below
>openssl req -x509 -newkey rsa:2048 -sha256 -days 365 -keyout example.key -out example.crt -subj "/CN=example.com"

But it errored out as seen above, could it be a misconfiguration in my machine or how I installed `openssl`?

[the remark this answers is not in the document]
Comment (derisen) [reply lib/msal-node/docs/certificate-credentials.md:23]: huh 😕 looks like there's a [similar issue here](https://github.com/openssl/openssl/issues/8795). Does any of the workarounds they discuss there help?

**Comment by mmacy** [comment]
@derisen For all the cert creation and upload-to-Key Vault stuff, would it be possible to use the Azure CLI (in the [Azure Cloud Shell](https://docs.microsoft.com/en-us/azure/cloud-shell/overview)) instead of doing it all manually?

[az ad sp create-for-rbac](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest#az_ad_sp_create_for_rbac)

```Bash
# Create using a self-signed certificate, and store it within KeyVault.
az ad sp create-for-rbac --keyvault MyVault --cert CertName --create-cert
```

image

**Comment by derisen** [comment]
> @derisen For all the cert creation and upload-to-Key Vault stuff, would it be possible to use the Azure CLI (in the [Azure Cloud Shell](https://docs.microsoft.com/en-us/azure/cloud-shell/overview)) instead of doing it all manually?
> 
> [az ad sp create-for-rbac](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest#az_ad_sp_create_for_rbac)
> 
> ```shell
> # Create using a self-signed certificate, and store it within KeyVault.
> az ad sp create-for-rbac --keyvault MyVault --cert CertName --create-cert
> ```
> 
> image

@mmacy I can absolutely check this out and add it as a section or an alternative (to the second doc perhaps, _key-vault-managed-identity.md_. I'm reluctant to replace cert generation with OpenSSL completely, as it is an industry standard and works with all platforms

File: lib/msal-node/docs/key-vault-managed-identity.md (hunk 2 above)
Comment (sadasant) [reply lib/msal-node/docs/key-vault-managed-identity.md:80]: Thank you, @derisen

File: lib/msal-node/docs/certificate-credentials.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (samuelkubai) [reply lib/msal-node/docs/certificate-credentials.md:23]: Information in the similar issue fixed mine. Marking this as resolved.

**Review by sameerag — APPROVED** [review]

**Pull request merged**